### PR TITLE
#185233328 Integrate Categories Section with Backend

### PR DESCRIPTION
**Pull request by AbdulKhaliq59**
#### What does this PR do?
This PR is used to integrate categories Section with backend
#### Description of Task to be completed?
- [x] Filter and display all categories that are in all products
- [x] Display that products related to category accordingly
- [x] Test the functionality with jest
#### How should this be manually tested?
- `run GET: /`
#### What are the relevant pivotal tracker/Trello stories?
Story ID: [#185233328](https://www.pivotaltracker.com/story/show/185233328)
#### Screenshots (if appropriate)
screencapture-localhost-3000-2023-05-25-10_29_33

#### Questions:

### CODE REVIEW

File: src/components/TopCategories/index.jsx [hunk 1]
```diff
@@ -1,46 +1,92 @@
...
+	useEffect(() => {
+		const fetchAllProducts = async () => {
+			let allProducts = [];
+			let currentPage = 1;
+			let totalPages = 1;
+
+			while (currentPage <= totalPages) {
+				const response = await fetch(
+					`https://e-comm-team-amigos-bn-project.onrender.com/product?page=${currentPage}`,
+					{
```
Comment (GasanaJr) [inline src/components/TopCategories/index.jsx]: Fetching should be done in redux

File: src/components/TodayDeal/index.jsx [hunk 2]
```diff
@@ -1,89 +1,148 @@
...
+			setIsLoading(false);
+		};
+
+		fetchAllProducts();
+	}, []);
+	const handleCategorySearch = async (category) => {
+		{
+			const response = await fetch(
+				`https://e-comm-team-amigos-bn-project.onrender.com/product/search?category=${encodeURIComponent(
+					category,
```
Comment (GasanaJr) [inline src/components/TodayDeal/index.jsx]: Fetching should be done in redux

File: src/components/TodayDeal/index.jsx [hunk 3]
```diff
@@ -1,89 +1,148 @@
...
+		const fetchAllProducts = async () => {
+			setIsLoading(true);
+			let allProducts = [];
+			let currentPage = 1;
+			let totalPages = 1;
+
+			while (currentPage <= totalPages) {
+				const response = await fetch(
+					`https://e-comm-team-amigos-bn-project.onrender.com/product?page=${currentPage}`,
+					{
```
Comment (GasanaJr) [inline src/components/TodayDeal/index.jsx]: Use redux

**Review by GasanaJr — CHANGES_REQUESTED** [review]
Address the comments I left

**Comment by khevin2** [comment]
Please use Env variable and replace those hard coded backed links

**Review by Emile-SHUMBUSHO — COMMENTED** [review]
When selecting categories, there should be a loading indicator to show a user that there is a fetching of data that is going on, otherwise you have did a great job.

**Review by wilbrord2 — APPROVED** [review]

File: src/redux/actions/homeAction.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+				];
+				currentPage++;
+				totalPages = data.responseData.totalPages;
+			}
+
+			// Extract unique categories from the products
+			const uniqueCategories = Array.from(
+				new Set(allProducts.map((product) => product.category)),
+			);
+			console.log(uniqueCategories);
```
Comment (GasanaJr) [inline src/redux/actions/homeAction.js]: the console logs are not necessary here

**Review by GasanaJr — CHANGES_REQUESTED** [review]
address the comment I left

**Review by ericndungutse — APPROVED** [review]
LGTM

**Review by Emile-SHUMBUSHO — APPROVED** [review]

**Review by GasanaJr — APPROVED** [review]

File: src/components/TopCategories/index.jsx [hunk 5]
```diff
@@ -1,46 +1,92 @@
...
+						headers: {
+							"Content-Type": "application/json",
+						},
+					},
```
Comment (Habinezajanvier) [inline src/components/TopCategories/index.jsx]: Make sure you are not fetching inside components

**Review by Habinezajanvier — CHANGES_REQUESTED** [review]

**Review by Mystica52 — APPROVED** [review]

**Review by Habinezajanvier — APPROVED** [review]
👍

**Pull request merged**